lib/vault.ts

```diff
@@ -184,3 +311,125 @@ export async function numberOfUnprocessedSubmissions(
+  ability: UserAbility,
+  dynamoDb: DynamoDBDocumentClient,
+  formResponses: NonNullable<VaultSubmissionAndConfirmationList>[],
+  formID: string
+): Promise<{ responsesDeleted: number }> {
+  try {
+    await checkAbilityToDeleteSubmissions(ability, formID);
+  } catch (e) {
+    if (e instanceof AccessControlError)
+      logEvent(
+        ability.userID,
+        { type: "Response" },
+        "AccessDenied",
+        `Attempted to delete all responses for form ${formID}`
+      );
+    throw e;
+  }
+
+  let responsesDeleted = 0;
+  /**
+   * The `BatchWriteItemCommand` can only take up to 25 `DeleteRequest` at a time.
+   * We have to delete 2 items from DynamoDB for each form response (12*2=24).
+   */
+  logMessage.debug(`----start list form response----`);
+  logMessage.debug(formResponses);
+  logMessage.debug(`----end list form response----`);
+
+  for (const formResponsesChunk of chunkArray(formResponses, 12)) {
+    const deleteRequests = formResponsesChunk.flatMap(
+      (formResponse: VaultSubmissionAndConfirmationList) => {
+        return [
+          {
+            DeleteRequest: {
+              Key: {
+                FormID: {
+                  S: formResponse.formID,
+                },
+                NAME_OR_CONF: {
+                  S: `NAME#${formResponse.name}`,
+                },
+              },
+            },
+          },
+          {
+            DeleteRequest: {
+              Key: {
+                FormID: {
+                  S: formResponse.formID,
+                },
+                NAME_OR_CONF: {
+                  S: `CONF#${formResponse.confirmationCode}`,
+                },
+              },
+            },
+          },
+        ];
+      }
+    );
+
+    logMessage.debug(`Deleting ${deleteRequests.length} form responses from DynamoDB.`);
+    logMessage.debug(deleteRequests);
+    logMessage.debug(`Delete requests length ${deleteRequests.length}`);
+    logMessage.debug(deleteRequests.length / 2);
+
+    const batchWriteItemCommandInput = {
+      RequestItems: {
+        ["Vault"]: deleteRequests,
+      },
+    };
+
+    let response;
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      response = await dynamoDb.send(new BatchWriteItemCommand(batchWriteItemCommandInput));
+      logMessage.debug(`------start response------`);
+      logMessage.debug(response);
+      responsesDeleted += deleteRequests.length / 2;
+      logMessage.debug(`Deleted responses ${responsesDeleted}`);
+      logMessage.debug(`------end response------`);
+    } catch (e) {
+      logMessage.error(e);
+      throw new Error(`Failed to delete form responses from DynamoDB.`);
+    }
+  }
+  return { responsesDeleted };
+}
+
+export async function deleteDraftFormTestResponses(ability: UserAbility, formID: string) {
+  try {
+    // TODO: Verify that the form status is not Published (must be draft)
+    // const templateWithAssociatedUsers = await _unprotectedGetTemplateWithAssociatedUsers(formID);
```

Comment:
Even though it will result in another call to the Database it would be a good idea to check.  Instead of using a more complex call to functions in `Templates` a simple prisma call to the db could be used instead.
For example:
```js
  const isPublished = await prisma.template
    .findUnique({
      where: {
        id: formID,
      },
      select: {
        isPublished: true
        },
      },
    })
    .catch((e) => prismaErrors(e, true));
```